code/modules/mob/living/silicon/silicon.dm

```diff
@@ -353,3 +354,26 @@
 	..()
 	if(cameraFollow)
 		cameraFollow = null
+
+/mob/living/silicon/Move(newloc, direct)
+	..(newloc,direct)
+	if (underdoor)
+		if (!(layer == UNDERDOOR))
+			//We must have used hide, or had our layer changed by something else. cancel this processing
+			underdoor = 0
+
+		underdoor = 0
+		for (var/obj/machinery/door/D in loc)
+			if (D.hashatch)
+				underdoor = 1
+				break
+
+		if (!underdoor)
+			spawn(3)//A slight delay to let us finish walking out from under the door
+				layer = initial(layer)

```

Comment:
I already did. The under_door proc doesn't change a drone's layer unless its on a layer that draws over the door (hide mode draws under)

A drone in hide mode that walks under doors will remain in hide mode and not be changed. A drone not in hide mode will have its layer temporarily changed, then set back to the original value when it moves out.